Archipelago.HollowKnight/DeathLinkSupport.cs

```diff
@@ -0,0 +1,336 @@
+                new List<string> {
+                    "@ was in the wrong place.",
+                    "@ mistimed their jump.",
+                    "@ forgot their floaties.",
+                    "What @ thought was H2O was H2SO4.",
+                    "@ wishes they could swim.",
+                    "@ used the wrong kind of dive.",
+                    "@ got into a fight with a pool of liquid and lost.",
+                }
+            },
+            {
+                999,  // Deaths in the dream realm
+                new List<string> {
+                    "@ dozed off for good.",
+                    "@ was caught sleeping on the job.",
+                    "@ sought dreams but found only nightmares.",
+                    "@ got lost in Limbo.",
+                    "Good night, @.",
+                    "@ is resting in pieces.",
+                    "@ exploded into a thousand pieces of essence.",
+                    "Hey, @, you're finally awake.",
+                }
+            },
+        };
+    };
+
+    public class DeathLinkSupport
+    {
+        public readonly static DeathLinkSupport Instance = new();
+        private DeathLinkService service = null;
+        public bool Enabled { get; private set; } = false;
+        private DeathLinkType Mode { get => Archipelago.Instance.SlotOptions.DeathLink; }
+        private readonly System.Random random = new();  // This is only messaging, so does not need to be seeded.
+
+        private DeathLinkStatus Status;
+        private int outgoingDeathlinks;
+        private int lastDamageType;
+        private DateTime lastDamageTime;
+        private ShadeInfo shade;
+
+        private DeathLinkSupport()
+        {
+            Reset();
+        }
+
+        private void Reset()
+        {
+            lastDamageType = 0;
+            lastDamageTime = DateTime.MinValue;
+            outgoingDeathlinks = 0;
+            Status = DeathLinkStatus.None;
+            shade = null;
+        }
+        public void Enable()
+        {
+            var ap = Archipelago.Instance;
+            if (Enabled)
+            {
+                ap.LogWarn("Tried to enable already-enabled DeathlinkSupport; disabling first.");
+                Disable();
+            }
+            Enabled = true;
+            Reset();
+
+            ap.LogDebug($"Enabling DeathLink support, type: {Mode}");
+            service = ap.session.CreateDeathLinkServiceAndEnable();
+            service.OnDeathLinkReceived += OnDeathLinkReceived;
+            ModHooks.HeroUpdateHook += ModHooks_HeroUpdateHook;
+            ModHooks.BeforePlayerDeadHook += SendDeathLink;
+            ModHooks.AfterPlayerDeadHook += ModHooks_AfterPlayerDeadHook;
+            On.HeroController.TakeDamage += HeroController_TakeDamage;
+            ItemChanger.Events.AddFsmEdit(new FsmID("Hero Death Anim"), FsmEdit);
+            // SetupDreamDeathFSMOverride(true);
+        }
+
+        public void Disable()
```

Comment:
`Disable()` doesn't disable DeathLink on the client level. (It doesn't update tags to remove "DeathLink" tag.)

Reply:
I'm torn on this one.

DeathLinkProvider in Multiclient.net should probably have some functionality for this, but doesn't currently.

Manually removing the tag when we're not manually adding the tag seems messy.

And Disable() is currently only called as part of the same process that's about to disconnect from AP anyways.

Reply:
>And Disable() is currently only called as part of the same process that's about to disconnect from AP anyways.

Unless the DeathLinkSupport is being re-enabled in which case it disables then enables. So I'm not sure that's great. 

Good feedback on multiclient.net, will have to get back to that and add that in maybe.